lib/Sema/CSSimplify.cpp

```diff
@@ -3455,6 +3455,73 @@ performMemberLookup(ConstraintKind constraintKind, DeclName memberName,
   return result;
 }
 
+static bool fixOptionalBaseMemberConstraint(ConstraintSystem &cs,
+                                            Type unwrappedBaseObjTy,
+                                            DeclName member, Type memberTy,
+                                            DeclContext *useDC,
+                                            FunctionRefKind functionRefKind,
+                                            ArrayRef<OverloadChoice> outerAlternatives,
+                                            ConstraintLocator *locator) {
+  // Don't try to fixup multiple levels of optionality, if we unwrapped and
+  // the result is still optional, fail.
+  if (unwrappedBaseObjTy->getOptionalObjectType())
+    return false;
+
+  if (!(cs.Options & ConstraintSystemFlags::AllowFixes))
+    return false;
+
+  auto contextTy = cs.getContextualType();
+  bool forceUnwrap;
+  if (contextTy)
+    forceUnwrap = contextTy->getOptionalObjectType().isNull();
+  else
+    forceUnwrap =
+        cs.Options.contains(ConstraintSystemFlags::PreferForceUnwrapToOptional);
+
+  if (forceUnwrap || functionRefKind != FunctionRefKind::Unapplied) {
+    // Note the fix.
+    auto fixKind =
+        forceUnwrap ? FixKind::ForceOptional : FixKind::OptionalChaining;
+    if (cs.recordFix(fixKind, locator))
+      return false;
+
+    // The simple case - just look through one level of optional.
+    cs.addValueMemberConstraint(unwrappedBaseObjTy, member, memberTy, useDC,
+                                functionRefKind, outerAlternatives, locator);
+    return true;
+  }
+
+  auto innerTV =
+      cs.createTypeVariable(locator, TVO_CanBindToLValue | TVO_CanBindToInOut);
+  Type optTy = cs.getTypeChecker().getOptionalType(SourceLoc(), innerTV);
+  if (!optTy)
+    return false;
+
+  // Add a disjunction with the three possibilities:
+  // 1) member is non-optional, result of expression adds optional to it ("?")
+  // 2) member is non-optional, result of expression is non-optional ("!")
+  // 3) member is optional, result of expression matches it ("?")
+  SmallVector<Constraint *, 3> optionalities;
+  auto optionalResult = Constraint::createFixed(cs, ConstraintKind::Conversion,
```

Comment:
It's not clear to me why you are using `ConstraintKind::Conversion` in these three constraints rather than `ConstraintKind::Equal`.

Wouldn't the expectation be that the types are equal?